backend/fms_core/viewsets.py

```diff
@@ -552,6 +552,42 @@ class SampleViewSet(viewsets.ModelViewSet, TemplateActionsMixin):
         }
     ]
 
+    def get_queryset(self):
+        container_barcode = self.request.query_params.get('container__barcode__recursive')
+        container_name = self.request.query_params.get('container__name__recursive')
+        recursive = container_barcode or container_name
+
+        if recursive:
+            containers = Container.objects.all()
+            if container_barcode:
+                containers = containers.filter(barcode=container_barcode)
+            if container_name:
+                containers = containers.filter(name=container_name)
+
+            container_ids = tuple(containers.values_list('id', flat=True))
+
+            if not container_ids:
+                container_ids = tuple([None])
+
+            parent_containers = Container.objects.raw('''WITH RECURSIVE parent(id, location_id) AS (
+                                                               SELECT id, location_id
+                                                               FROM fms_core_container
+                                                               WHERE id IN %s
+
+                                                               UNION ALL
+
+                                                               SELECT child.id, child.location_id
+                                                               FROM fms_core_container AS child, parent
+                                                               WHERE child.location_id = parent.id
+                                                           )
+                                                           SELECT * FROM parent''', params=[container_ids])
+
+            return self.queryset.filter(container__in=parent_containers)
```

Comment:
I would set the self.queryset to the new value instead of having a second return statement.

Reply:
What do you mean?

Reply:
i would set the viewset internal queryset to the correct one instead of just returning it. In case somewhere else django actually uses it for one thing or another without going through get queryset. More than a single return statement in a function need to be justified in some way in my eyes also (readability, performance, ...).

Reply:
> More than a single return statement in a function need to be justified in some way

I have the opposite opinion, but I feel the the early-return style is more readable. E.g., first is the current style, second is early-return style:

```python
if a:
  # ...
  return a
return b
```

```python
if not a:
  return b
# ...
return b
```

It allows to filter out easy cases early, and the end of the function is where the hard cases are handled. Otherwise you need to jump many lines of code to make the connection between the `if` and the `return`, and that is harder to understand.